### PR TITLE
Support CMA-ES with Margin.

### PR DESCRIPTION
**Pull request by knshnb**
The implementation of CMA-ES with margin by #115 has the following issues.
- It is implicit that discrete spaces must be located at the latter dimensions.
- The argument of `__init__` is inconsistent with `CMA`.

This PR is a continuation of #115 and changes the interface of `CMAwM` to address the issues. The example code would be changed as follows.
```python3
import numpy as np
from cmaes import CMAwM


def ellipsoid_onemax(x, n_zdim):
    n = len(x)
    n_rdim = n - n_zdim
    r = 10
    if len(x) < 2:
        raise ValueError("dimension must be greater one")
    ellipsoid = sum([(1000 ** (i / (n_rdim - 1)) * x[i]) ** 2 for i in range(n_rdim)])
    onemax = n_zdim - (0.0 < x[(n - n_zdim) :]).sum()
    return ellipsoid + r * onemax


def main():
    binary_dim, continuous_dim = 10, 10
    dim = binary_dim + continuous_dim
    bounds = np.concatenate(
        [
            np.tile([0, 1], (binary_dim, 1)),
            np.tile([-np.inf, np.inf], (continuous_dim, 1)),
        ]
    )
    steps = np.concatenate([np.ones(binary_dim), np.zeros(continuous_dim)])
    optimizer = CMAwM(mean=np.zeros(dim), sigma=2.0, bounds=bounds, steps=steps)
    print(" evals    f(x)")
    print("======  ==========")

    evals = 0
    while True:
        solutions = []
        for _ in range(optimizer.population_size):
            x_for_eval, x_for_tell = optimizer.ask()
            value = ellipsoid_onemax(x_for_eval, binary_dim)
            evals += 1
            solutions.append((x_for_tell, value))
            if evals % 300 == 0:
                print(f"{evals:5d}  {value:10.5f}")
        optimizer.tell(solutions)

        if optimizer.should_stop():
            break


if __name__ == "__main__":
    main()
```

### CODE REVIEW

**Comment by c-bata** [comment]
Thank you for your pull request!

@nomuramasahir0 Could you trigger the GitHub action on this PR?

**Comment by c-bata** [comment]
I opened https://github.com/CyberAgentAILab/cmaes/pull/122 to fix CI problems :bow:

File: cmaes/_cmawm.py [hunk 1]
```diff
@@ -0,0 +1,642 @@
...
+        population_size: Optional[int] = None,
+        cov: Optional[np.ndarray] = None,
+        margin: Optional[float] = None,
+    ):
+        assert len(bounds) == len(steps), "bounds and steps must be the same length"
+        assert not np.isnan(steps).any(), "steps should not include NaN"
+
+        # split discrete space and continuous space
+        discrete_idx = np.where(steps > 0)[0]
+        continuous_idx = np.where(steps <= 0)[0]
```
Comment (c-bata) [inline cmaes/_cmawm.py:113]: nit

```suggestion
        continuous_idx = ~discrete_idx
```

File: cmaes/_cmawm.py [hunk 2]
```diff
@@ -0,0 +1,649 @@
+import math
+import numpy as np
+from scipy.stats import chi2
+from scipy.stats import norm
```
Comment (c-bata) [inline cmaes/_cmawm.py:4]: Should we add `scipy` in `extra_requires`?

File: cmaes/_cmawm.py [hunk 3]
```diff
@@ -0,0 +1,649 @@
...
+        seed: Optional[int] = None,
+        population_size: Optional[int] = None,
+        cov: Optional[np.ndarray] = None,
+        margin: Optional[float] = None,
+    ):
+        assert len(bounds) == len(steps), "bounds and steps must be the same length"
+        assert not np.isnan(steps).any(), "steps should not include NaN"
+
+        # split discrete space and continuous space
+        self._discrete_idx = np.where(steps > 0)[0]
```
Comment (c-bata) [inline cmaes/_cmawm.py:112]: Should we add an assertion to ensure the existence of discrete parameters?

**Review by c-bata — COMMENTED** [review]
Changes overall looks good to me and the current interface looks more straightforward than #115.
We might want to consider refactoring duplicated code with `CMA` class, but as for this change, the current implementation looks good to me. We should work on it as follow-up tasks.

I left two minor comments.

File: cmaes/_cmawm.py (hunk 2 above)
Comment (knshnb) [reply cmaes/_cmawm.py:4]: I added `cmawm = scipy`. ( https://github.com/CyberAgentAILab/cmaes/pull/121/commits/b3a960850d833d9c399f368c5afaab698cafcf5f)

File: cmaes/_cmawm.py (hunk 3 above)
Comment (knshnb) [reply cmaes/_cmawm.py:112]: Thanks for pointing it out! I added that validation. (https://github.com/CyberAgentAILab/cmaes/pull/121/commits/1ef697199592f2e2433fa3648c95ff3d24afd3b5)

**Comment by knshnb** [comment]
Thanks for the review! I applied your comments.

I'm sorry for the PR with a large change. Please see `cmaes/_cmawm.py` and `examples/cmaes_with_margin.py` in https://github.com/CyberAgentAILab/cmaes/compare/3938dee69ebb707cf68ad7aaebd299001dec0ce7..1ef697199592f2e2433fa3648c95ff3d24afd3b5 for the main change from #115.

Although the behavior with the same seed is different from #115 (because I changed the order of internal representations), I checked that we can obtain similar tendencies in benchmark problems. 2000 trials, 100 runs for each sampler in each problem:
all
(_CmaEsSampler_NopPruner_0: Original, _CmaEsSampler_NopPruner_1: CMA-ES with margin)

**Review by c-bata — APPROVED** [review]
Thanks for the update! LGTM!

**Pull request merged**